diff -urN --no-dereference gcr-clean/meson.build gcr-workdir/meson.build
--- gcr-clean/meson.build
+++ gcr-workdir/meson.build
@@ -52,8 +52,14 @@
   error('Couldn\'t find location for pkcs11 module config')
 endif
 libsecret_dep = dependency('libsecret-1', version: '>= 0.20', required: get_option('ssh_agent'))
-ssh_add_path = find_program('ssh-add', required: get_option('ssh_agent')).path()
-ssh_agent_path = find_program('ssh-agent', required: get_option('ssh_agent')).path()
+ssh_add_path = get_option('ssh_add_path')
+if ssh_add_path == ''
+  ssh_add_path = find_program('ssh-add', required: get_option('ssh_agent')).path()
+endif
+ssh_agent_path = get_option('ssh_agent_path')
+if ssh_agent_path == ''
+  ssh_agent_path = find_program('ssh-agent', required: get_option('ssh_agent')).path()
+endif
 
 with_systemd = false
 libsystemd_deps = []
diff -urN --no-dereference gcr-clean/meson_options.txt gcr-workdir/meson_options.txt
--- gcr-clean/meson_options.txt
+++ gcr-workdir/meson_options.txt
@@ -18,6 +18,16 @@
   value: '',
   description: 'Path to gpg, autodetected if not set',
 )
+option('ssh_add_path',
+  type: 'string',
+  value: '',
+  description: 'Path to ssh-add, autodetected if not set',
+)
+option('ssh_agent_path',
+  type: 'string',
+  value: '',
+  description: 'Path to ssh-agent, autodetected if not set',
+)
 option('ssh_agent',
   type: 'boolean',
   value: true,
