diff --git rust-clean/Cargo.lock rust-workdir/Cargo.lock
index a170ece..4c2b542 100644
--- rust-clean/Cargo.lock
+++ rust-workdir/Cargo.lock
@@ -2045,9 +2045,7 @@ checksum = "9fa0e2a1fcbe2f6be6c42e342259976206b383122fc152e872795338b5a3f3a7"
 
 [[package]]
 name = "libc"
-version = "0.2.174"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "1171693293099992e19cddea4e8b849964e9846f4acee11b3948bcc337be8776"
+version = "0.2.172"
 
 [[package]]
 name = "libdbus-sys"
diff --git rust-clean/Cargo.toml rust-workdir/Cargo.toml
index e08f14d..ac853c0 100644
--- rust-clean/Cargo.toml
+++ rust-workdir/Cargo.toml
@@ -89,3 +89,6 @@ codegen-units = 1
 # FIXME: LTO cannot be enabled for binaries in a workspace
 # <https://github.com/rust-lang/cargo/issues/9330>
 # lto = true
+
+[patch.crates-io]
+libc = { path = "/base_dir/sources/rust-libc" }
diff --git rust-workdir/compiler/rustc_target/src/spec/base/linux_mlibc.rs rust-workdir/compiler/rustc_target/src/spec/base/linux_mlibc.rs
new file mode 100644
index 0000000..2c79ea2
--- /dev/null
+++ rust-workdir/compiler/rustc_target/src/spec/base/linux_mlibc.rs
@@ -0,0 +1,27 @@
+use crate::spec::{base, Cc, LinkArgs, LinkerFlavor, Lld, TargetOptions};
+
+pub(crate) fn opts() -> TargetOptions {
+    let mut base = base::linux::opts();
+    let mut args = LinkArgs::new();
+
+    args.insert(
+        LinkerFlavor::Gnu(Cc::Yes, Lld::No),
+        vec![
+            // We want to be able to strip as much executable code as possible
+            // from the linker command line, and this flag indicates to the
+            // linker that it can avoid linking in dynamic libraries that don't
+            // actually satisfy any symbols up to that point (as with many other
+            // resolutions the linker does). This option only applies to all
+            // following libraries so we're sure to pass it as one of the first
+            // arguments.
+            "-Wl,--as-needed".to_string().into(),
+            // Always enable NX protection when it is available
+            "-Wl,-z,noexecstack".to_string().into(),
+        ],
+    );
+
+    base.env = "mlibc".into();
+    base.pre_link_args = args;
+
+    base
+}
diff --git rust-clean/compiler/rustc_target/src/spec/base/mod.rs rust-workdir/compiler/rustc_target/src/spec/base/mod.rs
index b368d93..5933ce5 100644
--- rust-clean/compiler/rustc_target/src/spec/base/mod.rs
+++ rust-workdir/compiler/rustc_target/src/spec/base/mod.rs
@@ -15,6 +15,7 @@
 pub(crate) mod l4re;
 pub(crate) mod linux;
 pub(crate) mod linux_gnu;
+pub(crate) mod linux_mlibc;
 pub(crate) mod linux_musl;
 pub(crate) mod linux_ohos;
 pub(crate) mod linux_uclibc;
diff --git rust-clean/compiler/rustc_target/src/spec/mod.rs rust-workdir/compiler/rustc_target/src/spec/mod.rs
index 7a49f00..aea77b4 100644
--- rust-clean/compiler/rustc_target/src/spec/mod.rs
+++ rust-workdir/compiler/rustc_target/src/spec/mod.rs
@@ -1747,6 +1747,7 @@ fn $module() {
 
 supported_targets! {
     ("x86_64-unknown-linux-gnu", x86_64_unknown_linux_gnu),
+    ("x86_64-unknown-linux-mlibc", x86_64_unknown_linux_mlibc),
     ("x86_64-unknown-linux-gnux32", x86_64_unknown_linux_gnux32),
     ("i686-unknown-linux-gnu", i686_unknown_linux_gnu),
     ("i586-unknown-linux-gnu", i586_unknown_linux_gnu),
diff --git rust-workdir/compiler/rustc_target/src/spec/targets/x86_64_unknown_linux_mlibc.rs rust-workdir/compiler/rustc_target/src/spec/targets/x86_64_unknown_linux_mlibc.rs
new file mode 100644
index 0000000..62f38e9
--- /dev/null
+++ rust-workdir/compiler/rustc_target/src/spec/targets/x86_64_unknown_linux_mlibc.rs
@@ -0,0 +1,27 @@
+use crate::spec::{base, Cc, LinkerFlavor, Lld, StackProbeType, Target};
+
+pub(crate) fn target() -> Target {
+    let mut base = base::linux_mlibc::opts();
+    base.cpu = "x86-64".into();
+    base.max_atomic_width = Some(64);
+    base.pre_link_args
+        .get_mut(&LinkerFlavor::Gnu(Cc::Yes, Lld::No))
+        .unwrap()
+        .push("-m64".into());
+    base.stack_probes = StackProbeType::Inline;
+
+    Target {
+        llvm_target: "x86_64-unknown-linux-mlibc".into(),
+        metadata: crate::spec::TargetMetadata {
+            description: Some("64-bit Linux with mlibc".into()),
+            tier: Some(3),
+            host_tools: Some(true),
+            std: Some(true),
+        },
+        pointer_width: 64,
+        data_layout:
+            "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128".into(),
+        arch: "x86_64".into(),
+        options: base,
+    }
+}
diff --git rust-clean/library/Cargo.lock rust-workdir/library/Cargo.lock
index 4f0e7ae..363e822 100644
--- rust-clean/library/Cargo.lock
+++ rust-workdir/library/Cargo.lock
@@ -141,9 +141,7 @@ dependencies = [
 
 [[package]]
 name = "libc"
-version = "0.2.174"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "1171693293099992e19cddea4e8b849964e9846f4acee11b3948bcc337be8776"
+version = "0.2.172"
 dependencies = [
  "rustc-std-workspace-core",
 ]
diff --git rust-clean/library/Cargo.toml rust-workdir/library/Cargo.toml
index c66f621..73bd379 100644
--- rust-clean/library/Cargo.toml
+++ rust-workdir/library/Cargo.toml
@@ -57,6 +57,7 @@ rustflags = ["-Cpanic=abort"]
 [patch.crates-io]
 # See comments in `library/rustc-std-workspace-core/README.md` for what's going on
 # here
+libc = { path = "/base_dir/sources/rust-libc" }
 rustc-std-workspace-core = { path = 'rustc-std-workspace-core' }
 rustc-std-workspace-alloc = { path = 'rustc-std-workspace-alloc' }
 rustc-std-workspace-std = { path = 'rustc-std-workspace-std' }
diff --git rust-clean/library/std/src/sys/fs/unix.rs rust-workdir/library/std/src/sys/fs/unix.rs
index dc27827..b45c52b 100644
--- rust-clean/library/std/src/sys/fs/unix.rs
+++ rust-workdir/library/std/src/sys/fs/unix.rs
@@ -18,7 +18,7 @@
 use libc::dirfd;
 #[cfg(any(target_os = "fuchsia", target_os = "illumos"))]
 use libc::fstatat as fstatat64;
-#[cfg(any(all(target_os = "linux", not(target_env = "musl")), target_os = "hurd"))]
+#[cfg(any(all(target_os = "linux", not(any(target_env = "musl", target_env = "mlibc"))), target_os = "hurd"))]
 use libc::fstatat64;
 #[cfg(any(
     target_os = "android",
@@ -29,7 +29,7 @@
     target_os = "aix",
     target_os = "nto",
     target_os = "vita",
-    all(target_os = "linux", target_env = "musl"),
+    all(target_os = "linux", any(target_env = "musl", target_env = "mlibc")),
 ))]
 use libc::readdir as readdir64;
 #[cfg(not(any(
@@ -46,7 +46,7 @@
     target_os = "hurd",
 )))]
 use libc::readdir_r as readdir64_r;
-#[cfg(any(all(target_os = "linux", not(target_env = "musl")), target_os = "hurd"))]
+#[cfg(any(all(target_os = "linux", not(any(target_env = "musl", target_env = "mlibc"))), target_os = "hurd"))]
 use libc::readdir64;
 #[cfg(target_os = "l4re")]
 use libc::readdir64_r;
@@ -57,7 +57,7 @@
     lstat as lstat64, off64_t, open as open64, stat as stat64,
 };
 #[cfg(not(any(
-    all(target_os = "linux", not(target_env = "musl")),
+    all(target_os = "linux", not(any(target_env = "musl", target_env = "mlibc"))),
     target_os = "l4re",
     target_os = "android",
     target_os = "hurd",
@@ -67,7 +67,7 @@
     lstat as lstat64, off_t as off64_t, open as open64, stat as stat64,
 };
 #[cfg(any(
-    all(target_os = "linux", not(target_env = "musl")),
+    all(target_os = "linux", not(any(target_env = "musl", target_env = "mlibc"))),
     target_os = "l4re",
     target_os = "hurd"
 ))]
@@ -891,7 +891,7 @@ pub fn file_name(&self) -> OsString {
 
     #[cfg(all(
         any(
-            all(target_os = "linux", not(target_env = "musl")),
+            all(target_os = "linux", not(any(target_env = "musl", target_env = "mlibc"))),
             target_os = "android",
             target_os = "fuchsia",
             target_os = "hurd",
@@ -921,7 +921,7 @@ pub fn metadata(&self) -> io::Result<FileAttr> {
 
     #[cfg(any(
         not(any(
-            all(target_os = "linux", not(target_env = "musl")),
+            all(target_os = "linux", not(any(target_env = "musl", target_env = "mlibc"))),
             target_os = "android",
             target_os = "fuchsia",
             target_os = "hurd",
diff --git rust-clean/library/std/src/sys/pal/unix/futex.rs rust-workdir/library/std/src/sys/pal/unix/futex.rs
index c23278b..d18c696 100644
--- rust-clean/library/std/src/sys/pal/unix/futex.rs
+++ rust-workdir/library/std/src/sys/pal/unix/futex.rs
@@ -21,6 +21,58 @@
 /// Must be the underlying type of SmallFutex
 pub type SmallPrimitive = u32;
 
+#[cfg(target_env = "mlibc")]
+unsafe fn sys_futex_wait(
+    futex: *const Atomic<u32>,
+    op: i32,
+    expected: u32,
+    timeout: *const libc::timespec,
+    uaddr2: *const u32,
+    val3: u32,
+) -> core::ffi::c_long {
+    let r = libc::__do_syscall6(
+        libc::SYS_futex,
+        futex as _,
+        op as _,
+        expected as _,
+        timeout as _,
+        uaddr2 as _,
+        val3 as _,
+    );
+    libc::__do_syscall_ret(r)
+}
+
+#[cfg(not(target_env = "mlibc"))]
+unsafe fn sys_futex_wait(
+    futex: *const Atomic<u32>,
+    op: i32,
+    expected: u32,
+    timeout: *const libc::timespec,
+    uaddr2: *const u32,
+    val3: u32,
+) -> core::ffi::c_long {
+    libc::syscall(
+        libc::SYS_futex,
+        futex,
+        op,
+        expected,
+        timeout,
+        uaddr2, // This argument is unused for FUTEX_WAIT_BITSET.
+        val3,   // A full bitmask, to make it behave like a regular FUTEX_WAIT.
+    )
+}
+
+#[cfg(target_env = "mlibc")]
+unsafe fn sys_futex_wake(futex: *const Atomic<u32>, op: i32, count: i32) -> bool {
+    let r = libc::__do_syscall3(libc::SYS_futex, futex as _, op as _, count as _);
+    libc::__do_syscall_ret(r) > 0
+}
+
+#[cfg(not(target_env = "mlibc"))]
+unsafe fn sys_futex_wake(futex: *const Atomic<u32>, op: i32, count: i32) -> bool {
+    libc::syscall(libc::SYS_futex, futex, op, count) > 0
+}
+
 /// Waits for a `futex_wake` operation to wake us.
 ///
 /// Returns directly if the futex doesn't hold the expected value.
@@ -69,8 +121,7 @@ pub fn futex_wait(futex: &Atomic<u32>, expected: u32, timeout: Option<Duration>)
                 } else if #[cfg(any(target_os = "linux", target_os = "android"))] {
                     // Use FUTEX_WAIT_BITSET rather than FUTEX_WAIT to be able to give an
                     // absolute time rather than a relative time.
-                    libc::syscall(
-                        libc::SYS_futex,
+                    sys_futex_wait(
                         futex as *const Atomic<u32>,
                         libc::FUTEX_WAIT_BITSET | libc::FUTEX_PRIVATE_FLAG,
                         expected,
@@ -100,18 +151,24 @@ pub fn futex_wait(futex: &Atomic<u32>, expected: u32, timeout: Option<Duration>)
 /// On some platforms, this always returns false.
 #[cfg(any(target_os = "linux", target_os = "android"))]
 pub fn futex_wake(futex: &Atomic<u32>) -> bool {
-    let ptr = futex as *const Atomic<u32>;
-    let op = libc::FUTEX_WAKE | libc::FUTEX_PRIVATE_FLAG;
-    unsafe { libc::syscall(libc::SYS_futex, ptr, op, 1) > 0 }
+    unsafe {
+        sys_futex_wake(
+            futex as *const Atomic<u32>,
+            libc::FUTEX_WAKE | libc::FUTEX_PRIVATE_FLAG,
+            1,
+        )
+    }
 }
 
 /// Wakes up all threads that are waiting on `futex_wait` on this futex.
 #[cfg(any(target_os = "linux", target_os = "android"))]
 pub fn futex_wake_all(futex: &Atomic<u32>) {
-    let ptr = futex as *const Atomic<u32>;
-    let op = libc::FUTEX_WAKE | libc::FUTEX_PRIVATE_FLAG;
-    unsafe {
-        libc::syscall(libc::SYS_futex, ptr, op, i32::MAX);
+     unsafe {
+        sys_futex_wake(
+            futex as *const Atomic<u32>,
+            libc::FUTEX_WAKE | libc::FUTEX_PRIVATE_FLAG,
+            i32::MAX,
+        );
     }
 }
 
diff --git rust-clean/library/std/src/sys/pal/unix/os.rs rust-workdir/library/std/src/sys/pal/unix/os.rs
index 850bdfd..fe62829 100644
--- rust-clean/library/std/src/sys/pal/unix/os.rs
+++ rust-workdir/library/std/src/sys/pal/unix/os.rs
@@ -128,10 +128,12 @@ pub fn error_string(errno: i32) -> String {
                     target_env = "newlib",
                     target_os = "cygwin"
                 ),
-                not(target_env = "ohos")
+                not(target_env = "ohos"),
+                not(target_env = "mlibc")
             ),
             link_name = "__xpg_strerror_r"
         )]
+        #[cfg_attr(target_env = "mlibc", link_name = "__gnu_strerror_r")]
         fn strerror_r(errnum: c_int, buf: *mut c_char, buflen: libc::size_t) -> c_int;
     }
 
diff --git rust-clean/library/std/src/sys/pal/unix/weak.rs rust-workdir/library/std/src/sys/pal/unix/weak.rs
index c8cf75b..7e8de04 100644
--- rust-clean/library/std/src/sys/pal/unix/weak.rs
+++ rust-workdir/library/std/src/sys/pal/unix/weak.rs
@@ -152,7 +152,7 @@ unsafe fn fetch(name: &str) -> *mut libc::c_void {
     unsafe { libc::dlsym(libc::RTLD_DEFAULT, name.as_ptr()) }
 }
 
-#[cfg(not(any(target_os = "linux", target_os = "android")))]
+#[cfg(any(not(any(target_os = "linux", target_os = "android")), target_env = "mlibc"))]
 pub(crate) macro syscall {
     (fn $name:ident($($param:ident : $t:ty),* $(,)?) -> $ret:ty;) => (
         unsafe fn $name($($param: $t),*) -> $ret {
@@ -168,7 +168,7 @@ unsafe fn $name($($param: $t),*) -> $ret {
     )
 }
 
-#[cfg(any(target_os = "linux", target_os = "android"))]
+#[cfg(all(any(target_os = "linux", target_os = "android"), not(target_env = "mlibc")))]
 pub(crate) macro syscall {
     (
         fn $name:ident($($param:ident : $t:ty),* $(,)?) -> $ret:ty;
diff --git rust-clean/library/std/src/sys/process/unix/unix.rs rust-workdir/library/std/src/sys/process/unix/unix.rs
index 1fe80c1..a7ce6d2 100644
--- rust-clean/library/std/src/sys/process/unix/unix.rs
+++ rust-workdir/library/std/src/sys/process/unix/unix.rs
@@ -821,7 +821,12 @@ fn send_pidfd(&self, sock: &crate::sys::net::Socket) {
         unsafe {
             let child_pid = libc::getpid();
             // pidfd_open sets CLOEXEC by default
+            #[cfg(not(target_env = "mlibc"))]
             let pidfd = libc::syscall(libc::SYS_pidfd_open, child_pid, 0);
+            #[cfg(target_env = "mlibc")]
+            let pidfd = libc::__do_syscall_ret(
+                libc::__do_syscall2(libc::SYS_pidfd_open, child_pid as _, 0)
+            );
 
             let fds: [c_int; 1] = [pidfd as RawFd];
 
diff --git rust-clean/library/unwind/src/lib.rs rust-workdir/library/unwind/src/lib.rs
index 5451a38..651907f 100644
--- rust-clean/library/unwind/src/lib.rs
+++ rust-workdir/library/unwind/src/lib.rs
@@ -69,6 +69,10 @@
     }
 }
 
+#[cfg(target_env = "mlibc")]
+#[link(name = "gcc_s")]
+unsafe extern "C" {}
+
 // This is the same as musl except that we default to using the system libunwind
 // instead of libgcc.
 #[cfg(target_env = "ohos")]
diff --git rust-clean/src/bootstrap/Cargo.lock rust-workdir/src/bootstrap/Cargo.lock
index a6ca699..5b8ac72 100644
--- rust-clean/src/bootstrap/Cargo.lock
+++ rust-workdir/src/bootstrap/Cargo.lock
@@ -90,9 +90,9 @@ dependencies = [
 
 [[package]]
 name = "cc"
-version = "1.2.23"
+version = "1.2.28"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "5f4ac86a9e5bc1e2b3449ab9d7d3a6a405e3d1bb28d7b9be8614f55846ae3766"
+checksum = "4ad45f4f74e4e20eaa392913b7b33a7091c87e59628f4dd27888205ad888843c"
 dependencies = [
  "shlex",
 ]
diff --git rust-clean/src/bootstrap/Cargo.toml rust-workdir/src/bootstrap/Cargo.toml
index 9785a30..d2a5462 100644
--- rust-clean/src/bootstrap/Cargo.toml
+++ rust-workdir/src/bootstrap/Cargo.toml
@@ -32,7 +32,7 @@ test = false
 # Most of the time updating these dependencies requires modifications to the
 # bootstrap codebase(e.g., https://github.com/rust-lang/rust/issues/124565);
 # otherwise, some targets will fail. That's why these dependencies are explicitly pinned.
-cc = "=1.2.23"
+cc = "=1.2.28"
 cmake = "=0.1.54"
 
 build_helper = { path = "../build_helper" }
diff --git rust-clean/src/tools/build-manifest/src/main.rs rust-workdir/src/tools/build-manifest/src/main.rs
index 4c53ea4..260496d 100644
--- rust-clean/src/tools/build-manifest/src/main.rs
+++ rust-workdir/src/tools/build-manifest/src/main.rs
@@ -185,6 +185,7 @@
     "x86_64-unknown-illumos",
     "x86_64-unknown-linux-gnu",
     "x86_64-unknown-linux-gnux32",
+    "x86_64-unknown-linux-mlibc",
     "x86_64-unknown-linux-musl",
     "x86_64-unknown-linux-ohos",
     "x86_64-unknown-netbsd",
diff --git rust-clean/src/tools/generate-copyright/src/cargo_metadata.rs rust-workdir/src/tools/generate-copyright/src/cargo_metadata.rs
index 3fae26b..facaa16 100644
--- rust-clean/src/tools/generate-copyright/src/cargo_metadata.rs
+++ rust-workdir/src/tools/generate-copyright/src/cargo_metadata.rs
@@ -119,7 +119,11 @@ fn load_important_files(
 ) -> Result<(), Error> {
     let name_version = format!("{}-{}", package.name, package.version);
     println!("Scraping notices for {}...", name_version);
-    let dep_vendor_path = vendor_root.join(name_version);
+    let dep_vendor_path = vendor_root.join(&name_version);
+    if !dep_vendor_path.exists() {
+        println!("No vendor path for {} found, skipping", name_version);
+        return Ok(());
+    }
     for entry in std::fs::read_dir(dep_vendor_path)? {
         let entry = entry?;
         let metadata = entry.metadata()?;
