diff --git p11-kit-clean/trust/module.c p11-kit-workdir/trust/module.c
index f252516..a521877 100644
--- p11-kit-clean/trust/module.c
+++ p11-kit-workdir/trust/module.c
@@ -412,6 +412,10 @@ sys_C_Initialize (CK_VOID_PTR init_args)
 			                            p11_dict_ulongptr_equal,
 			                            NULL, p11_session_free);
 
+			char *paths = getenv ("P11_KIT_TRUST_PATH");
+			if (paths)
+				gl.paths = strdup (paths);
+
 			gl.tokens = p11_array_new ((p11_destroyer)p11_token_free);
 			if (gl.tokens && !create_tokens_inlock (gl.tokens, gl.paths ? gl.paths : TRUST_PATHS))
 				gl.tokens = NULL;
