diff --git libdvdread-clean/src/dvd_reader.c libdvdread-workdir/src/dvd_reader.c
index c4d9641..e0fc2c2 100644
--- libdvdread-clean/src/dvd_reader.c
+++ libdvdread-workdir/src/dvd_reader.c
@@ -29,6 +29,7 @@
 #include <stdio.h>          /* fprintf */
 #include <errno.h>          /* errno, EIN* */
 #include <string.h>         /* memcpy, strlen */
+#include <strings.h>
 #include <unistd.h>         /* pclose */
 #include <limits.h>         /* PATH_MAX */
 #include <dirent.h>         /* opendir, readdir */
