diff --git xorg-server-clean/configure xorg-server-workdir/configure
index 490a8e2..6e3de5a 100755
--- xorg-server-clean/configure
+++ xorg-server-workdir/configure
@@ -29347,6 +29347,7 @@ printf "%s\n" "#define BASE_FONT_PATH \"$ac_define_dir\"" >>confdefs.h
   test "$exec_prefix_NONE" && exec_prefix=NONE
 
 dridriverdir=`$PKG_CONFIG --variable=dridriverdir dri`
+dridriverdir=$(printf '%s' "$dridriverdir" | sed -e 's@/sysroot@@g')
 
   prefix_NONE=
   exec_prefix_NONE=
