diff --git gobject-introspection-clean/giscanner/shlibs.py gobject-introspection-workdir/giscanner/shlibs.py
index 9f8ab5d..7b457a8 100644
--- gobject-introspection-clean/giscanner/shlibs.py
+++ gobject-introspection-workdir/giscanner/shlibs.py
@@ -97,8 +97,11 @@ def _resolve_non_libtool(options, binary, libraries):
             args.extend(libtool)
             args.append('--mode=execute')
         platform_system = platform.system()
+        ldd_wrapper_from_env = os.environ.get("GI_LDD_WRAPPER")
         if options.ldd_wrapper:
             args.extend([options.ldd_wrapper, binary.args[0]])
+        elif ldd_wrapper_from_env:
+            args.extend([ldd_wrapper_from_env, binary.args[0]])
         elif platform_system == 'Darwin':
             args.extend(['otool', '-L', binary.args[0]])
         else:
diff --git gobject-introspection-clean/m4/introspection.m4 gobject-introspection-workdir/m4/introspection.m4
index b0ccd68..c47d2b4 100644
--- gobject-introspection-clean/m4/introspection.m4
+++ gobject-introspection-workdir/m4/introspection.m4
@@ -102,11 +102,11 @@ m4_define([_GOBJECT_INTROSPECTION_CHECK_INTERNAL],
     INTROSPECTION_GIRDIR=
     INTROSPECTION_TYPELIBDIR=
     if test "x$found_introspection" = "xyes"; then
-       INTROSPECTION_SCANNER=$PKG_CONFIG_SYSROOT_DIR`$PKG_CONFIG --variable=g_ir_scanner gobject-introspection-1.0`
-       INTROSPECTION_COMPILER=$PKG_CONFIG_SYSROOT_DIR`$PKG_CONFIG --variable=g_ir_compiler gobject-introspection-1.0`
-       INTROSPECTION_GENERATE=$PKG_CONFIG_SYSROOT_DIR`$PKG_CONFIG --variable=g_ir_generate gobject-introspection-1.0`
-       INTROSPECTION_GIRDIR=`$PKG_CONFIG --define-variable=datadir="${_GI_EXP_DATADIR}" --variable=girdir gobject-introspection-1.0`
-       INTROSPECTION_TYPELIBDIR="$($PKG_CONFIG --define-variable=libdir="${_GI_EXP_LIBDIR}" --variable=typelibdir gobject-introspection-1.0)"
+       INTROSPECTION_SCANNER=$INTROSPECTION_SCANNER_ENV
+       INTROSPECTION_COMPILER=$INTROSPECTION_COMPILER_ENV
+       INTROSPECTION_GENERATE=$INTROSPECTION_GENERATE_ENV
+       INTROSPECTION_GIRDIR=$INTROSPECTION_GIRDIR_ENV
+       INTROSPECTION_TYPELIBDIR=$INTROSPECTION_TYPELIBDIR_ENV
        INTROSPECTION_CFLAGS=`$PKG_CONFIG --cflags gobject-introspection-1.0`
        INTROSPECTION_LIBS=`$PKG_CONFIG --libs gobject-introspection-1.0`
        INTROSPECTION_MAKEFILE=$PKG_CONFIG_SYSROOT_DIR`$PKG_CONFIG --variable=datadir gobject-introspection-1.0`/gobject-introspection-1.0/Makefile.introspection
