diff -urN --no-dereference iptables-clean/extensions/dscp_helper.c iptables-workdir/extensions/dscp_helper.c
--- iptables-clean/extensions/dscp_helper.c
+++ iptables-workdir/extensions/dscp_helper.c
@@ -11,6 +11,7 @@
 
 #include <stdio.h>
 #include <string.h>
+#include <strings.h>
 #include <xtables.h>
 
 
diff -urN --no-dereference iptables-clean/extensions/libip6t_REJECT.c iptables-workdir/extensions/libip6t_REJECT.c
--- iptables-clean/extensions/libip6t_REJECT.c
+++ iptables-workdir/extensions/libip6t_REJECT.c
@@ -7,6 +7,7 @@
  */
 #include <stdio.h>
 #include <string.h>
+#include <strings.h>
 #include <xtables.h>
 #include <linux/netfilter_ipv6/ip6t_REJECT.h>
 
diff -urN --no-dereference iptables-clean/extensions/libip6t_mh.c iptables-workdir/extensions/libip6t_mh.c
--- iptables-clean/extensions/libip6t_mh.c
+++ iptables-workdir/extensions/libip6t_mh.c
@@ -14,6 +14,7 @@
 #include <stdint.h>
 #include <stdio.h>
 #include <string.h>
+#include <strings.h>
 #include <stdlib.h>
 #include <xtables.h>
 #include <linux/netfilter_ipv6/ip6t_mh.h>
diff -urN --no-dereference iptables-clean/extensions/libipt_REJECT.c iptables-workdir/extensions/libipt_REJECT.c
--- iptables-clean/extensions/libipt_REJECT.c
+++ iptables-workdir/extensions/libipt_REJECT.c
@@ -4,6 +4,7 @@
  */
 #include <stdio.h>
 #include <string.h>
+#include <strings.h>
 #include <xtables.h>
 #include <linux/netfilter_ipv4/ipt_REJECT.h>
 #include <linux/version.h>
diff -urN --no-dereference iptables-clean/extensions/libxt_AUDIT.c iptables-workdir/extensions/libxt_AUDIT.c
--- iptables-clean/extensions/libxt_AUDIT.c
+++ iptables-workdir/extensions/libxt_AUDIT.c
@@ -7,6 +7,7 @@
  */
 #include <stdio.h>
 #include <string.h>
+#include <strings.h>
 #include <xtables.h>
 #include <linux/netfilter/xt_AUDIT.h>
 
diff -urN --no-dereference iptables-clean/extensions/libxt_CT.c iptables-workdir/extensions/libxt_CT.c
--- iptables-clean/extensions/libxt_CT.c
+++ iptables-workdir/extensions/libxt_CT.c
@@ -4,6 +4,7 @@
 
 #include <stdio.h>
 #include <string.h>
+#include <strings.h>
 #include <xtables.h>
 #include <linux/netfilter/nf_conntrack_common.h>
 #include <linux/netfilter/xt_CT.h>
diff -urN --no-dereference iptables-clean/extensions/libxt_HMARK.c iptables-workdir/extensions/libxt_HMARK.c
--- iptables-clean/extensions/libxt_HMARK.c
+++ iptables-workdir/extensions/libxt_HMARK.c
@@ -15,6 +15,7 @@
 #include <stdbool.h>
 #include <stdio.h>
 #include <string.h>
+#include <strings.h>
 
 #include "xtables.h"
 #include <linux/netfilter/xt_HMARK.h>
diff -urN --no-dereference iptables-clean/extensions/libxt_LED.c iptables-workdir/extensions/libxt_LED.c
--- iptables-clean/extensions/libxt_LED.c
+++ iptables-workdir/extensions/libxt_LED.c
@@ -11,6 +11,7 @@
  */
 #include <stdio.h>
 #include <string.h>
+#include <strings.h>
 #include <stdlib.h>
 #include <xtables.h>
 #include <linux/netfilter/xt_LED.h>
diff -urN --no-dereference iptables-clean/extensions/libxt_RATEEST.c iptables-workdir/extensions/libxt_RATEEST.c
--- iptables-clean/extensions/libxt_RATEEST.c
+++ iptables-workdir/extensions/libxt_RATEEST.c
@@ -4,6 +4,7 @@
 
 #include <stdio.h>
 #include <string.h>
+#include <strings.h>
 #include <stdlib.h>
 #include <math.h>
 
diff -urN --no-dereference iptables-clean/extensions/libxt_conntrack.c iptables-workdir/extensions/libxt_conntrack.c
--- iptables-clean/extensions/libxt_conntrack.c
+++ iptables-workdir/extensions/libxt_conntrack.c
@@ -11,6 +11,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <strings.h>
 #include <xtables.h>
 #include <linux/netfilter/xt_conntrack.h>
 #include <linux/netfilter/xt_state.h>
diff -urN --no-dereference iptables-clean/extensions/libxt_dccp.c iptables-workdir/extensions/libxt_dccp.c
--- iptables-clean/extensions/libxt_dccp.c
+++ iptables-workdir/extensions/libxt_dccp.c
@@ -8,6 +8,7 @@
 #include <stdint.h>
 #include <stdio.h>
 #include <string.h>
+#include <strings.h>
 #include <stdlib.h>
 #include <netdb.h>
 #include <arpa/inet.h>
diff -urN --no-dereference iptables-clean/extensions/libxt_icmp.h iptables-workdir/extensions/libxt_icmp.h
--- iptables-clean/extensions/libxt_icmp.h
+++ iptables-workdir/extensions/libxt_icmp.h
@@ -1,3 +1,5 @@
+#include <strings.h>
+
 static const struct xt_icmp_names {
 	const char *name;
 	uint8_t type;
diff -urN --no-dereference iptables-clean/extensions/libxt_ipvs.c iptables-workdir/extensions/libxt_ipvs.c
--- iptables-clean/extensions/libxt_ipvs.c
+++ iptables-workdir/extensions/libxt_ipvs.c
@@ -8,6 +8,7 @@
 #include <stdbool.h>
 #include <stdio.h>
 #include <string.h>
+#include <strings.h>
 #include <xtables.h>
 #include <linux/ip_vs.h>
 #include <linux/netfilter/xt_ipvs.h>
diff -urN --no-dereference iptables-clean/extensions/libxt_pkttype.c iptables-workdir/extensions/libxt_pkttype.c
--- iptables-clean/extensions/libxt_pkttype.c
+++ iptables-workdir/extensions/libxt_pkttype.c
@@ -6,6 +6,7 @@
  */
 #include <stdio.h>
 #include <string.h>
+#include <strings.h>
 #include <xtables.h>
 #include <linux/if_packet.h>
 #include <linux/netfilter/xt_pkttype.h>
diff -urN --no-dereference iptables-clean/extensions/libxt_rateest.c iptables-workdir/extensions/libxt_rateest.c
--- iptables-clean/extensions/libxt_rateest.c
+++ iptables-workdir/extensions/libxt_rateest.c
@@ -5,6 +5,7 @@
 #include <stdbool.h>
 #include <stdio.h>
 #include <string.h>
+#include <strings.h>
 #include <stdlib.h>
 #include <stddef.h>
 #include <getopt.h>
diff -urN --no-dereference iptables-clean/extensions/libxt_sctp.c iptables-workdir/extensions/libxt_sctp.c
--- iptables-clean/extensions/libxt_sctp.c
+++ iptables-workdir/extensions/libxt_sctp.c
@@ -11,6 +11,7 @@
 #include <stdbool.h>
 #include <stdio.h>
 #include <string.h>
+#include <strings.h>
 #include <stdlib.h>
 #include <getopt.h>
 #include <netdb.h>
diff -urN --no-dereference iptables-clean/extensions/libxt_tcp.c iptables-workdir/extensions/libxt_tcp.c
--- iptables-clean/extensions/libxt_tcp.c
+++ iptables-workdir/extensions/libxt_tcp.c
@@ -3,6 +3,7 @@
 #include <stdio.h>
 #include <netdb.h>
 #include <string.h>
+#include <strings.h>
 #include <stdlib.h>
 #include <getopt.h>
 #include <netinet/in.h>
diff -urN --no-dereference iptables-clean/libxtables/getethertype.c iptables-workdir/libxtables/getethertype.c
--- iptables-clean/libxtables/getethertype.c
+++ iptables-workdir/libxtables/getethertype.c
@@ -40,6 +40,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <strings.h>
 #include <netinet/ether.h>
 #include <net/ethernet.h>
 #include <xtables.h>
diff -urN --no-dereference iptables-clean/libxtables/xtables.c iptables-workdir/libxtables/xtables.c
--- iptables-clean/libxtables/xtables.c
+++ iptables-workdir/libxtables/xtables.c
@@ -27,6 +27,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <strings.h>
 #include <unistd.h>
 #include <netinet/ether.h>
 #include <sys/socket.h>
diff -urN --no-dereference iptables-clean/libxtables/xtoptions.c iptables-workdir/libxtables/xtoptions.c
--- iptables-clean/libxtables/xtoptions.c
+++ iptables-workdir/libxtables/xtoptions.c
@@ -17,6 +17,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <strings.h>
 #include <syslog.h>
 #include <arpa/inet.h>
 #include <netinet/ip.h>
