diff -urN --no-dereference librsvg-clean/meson.build librsvg-workdir/meson.build
--- librsvg-clean/meson.build
+++ librsvg-workdir/meson.build
@@ -134,6 +134,9 @@
 
 # Extra env to pass to cargo
 extra_env = environment()
+extra_env.set('SYSTEM_DEPS_LIBRSVG_NO_PKG_CONFIG', '1')
+extra_env.set('SYSTEM_DEPS_LIBRSVG_SEARCH_NATIVE', meson.project_build_root() / 'rsvg')
+extra_env.set('SYSTEM_DEPS_LIBRSVG_LIB', 'rsvg-@0@'.format(librsvg_api_major))
 
 # If FreeType and/or libxml2 is/are found by CMake instead of
 # pkg-config, we must tell Cargo the libraries explicitly and
