diff -urN --no-dereference systemd-clean/src/basic/build-path.c systemd-workdir/src/basic/build-path.c
--- systemd-clean/src/basic/build-path.c
+++ systemd-workdir/src/basic/build-path.c
@@ -37,7 +37,7 @@
                          * using the bias calculated earlier. */
                         if (d->d_un.d_val != 0)
                                 strtab = (const char *) ((uintptr_t) d->d_un.d_val
-#if defined(__mips__) || defined(__riscv)
+#if defined(__mips__) || defined(__riscv) || defined(__mlibc__)
                                          + bias
 #endif
                                 );
diff -urN --no-dereference systemd-clean/src/core/exec-invoke.c systemd-workdir/src/core/exec-invoke.c
--- systemd-clean/src/core/exec-invoke.c
+++ systemd-workdir/src/core/exec-invoke.c
@@ -5,6 +5,7 @@
 #include <sys/ioctl.h>
 #include <sys/mount.h>
 #include <sys/prctl.h>
+#include <sys/file.h>
 
 #if HAVE_PAM
 #include <security/pam_appl.h>
diff -urN --no-dereference systemd-clean/src/firstboot/firstboot.c systemd-workdir/src/firstboot/firstboot.c
--- systemd-clean/src/firstboot/firstboot.c
+++ systemd-workdir/src/firstboot/firstboot.c
@@ -1,5 +1,6 @@
 /* SPDX-License-Identifier: LGPL-2.1-or-later */
 
+#include <sys/file.h>
 #include <fcntl.h>
 #include <getopt.h>
 #include <linux/loop.h>
diff -urN --no-dereference systemd-clean/src/libsystemd/sd-event/sd-event.c systemd-workdir/src/libsystemd/sd-event/sd-event.c
--- systemd-clean/src/libsystemd/sd-event/sd-event.c
+++ systemd-workdir/src/libsystemd/sd-event/sd-event.c
@@ -1891,7 +1891,8 @@
 
         usec_t before_timestamp = now(CLOCK_MONOTONIC);
         hashmap_trim_pools();
-        r = malloc_trim(0);
+        // r = malloc_trim(0);
+        r = 0;
         usec_t after_timestamp = now(CLOCK_MONOTONIC);
 
         if (r > 0)
diff -urN --no-dereference systemd-clean/src/mountfsd/mountwork.c systemd-workdir/src/mountfsd/mountwork.c
--- systemd-clean/src/mountfsd/mountwork.c
+++ systemd-workdir/src/mountfsd/mountwork.c
@@ -23,6 +23,8 @@
 #include "varlink.h"
 #include "varlink-io.systemd.MountFileSystem.h"
 
+#include <sys/file.h>
+
 #define ITERATIONS_MAX 64U
 #define RUNTIME_MAX_USEC (5 * USEC_PER_MINUTE)
 #define PRESSURE_SLEEP_TIME_USEC (50 * USEC_PER_MSEC)
diff -urN --no-dereference systemd-clean/src/nsresourced/userns-registry.c systemd-workdir/src/nsresourced/userns-registry.c
--- systemd-clean/src/nsresourced/userns-registry.c
+++ systemd-workdir/src/nsresourced/userns-registry.c
@@ -13,6 +13,8 @@
 #include "user-util.h"
 #include "userns-registry.h"
 
+#include <sys/file.h>
+
 int userns_registry_open_fd(void) {
         int fd;
 
diff -urN --no-dereference systemd-clean/src/portable/portable.c systemd-workdir/src/portable/portable.c
--- systemd-clean/src/portable/portable.c
+++ systemd-workdir/src/portable/portable.c
@@ -1,6 +1,7 @@
 /* SPDX-License-Identifier: LGPL-2.1-or-later */
 
 #include <linux/loop.h>
+#include <sys/file.h>
 
 #include "sd-messages.h"
 
diff -urN --no-dereference systemd-clean/src/shared/async.c systemd-workdir/src/shared/async.c
--- systemd-clean/src/shared/async.c
+++ systemd-workdir/src/shared/async.c
@@ -86,6 +86,9 @@
 }
 
 int asynchronous_close(int fd) {
+#if defined(__mlibc__)
+        safe_close(fd);
+#else
         unsigned v;
         pid_t pid;
         int r;
@@ -129,6 +132,7 @@
                         if (waitpid(pid, NULL, __WCLONE) >= 0 || errno != EINTR)
                                 break;
         }
+#endif
 
         return -EBADF; /* return an invalidated fd */
 }
diff -urN --no-dereference systemd-clean/src/shared/bus-util.c systemd-workdir/src/shared/bus-util.c
--- systemd-clean/src/shared/bus-util.c
+++ systemd-workdir/src/shared/bus-util.c
@@ -764,9 +764,9 @@
         if (!f)
                 return -ENOMEM;
 
-        r = RET_NERRNO(malloc_info(/* options= */ 0, f));
-        if (r < 0)
-                return r;
+        // r = RET_NERRNO(malloc_info(/* options= */ 0, f));
+        // if (r < 0)
+        //         return r;
 
         r = memstream_finalize(&m, &dump, &dump_size);
         if (r < 0)
diff -urN --no-dereference systemd-clean/src/shared/common-signal.c systemd-workdir/src/shared/common-signal.c
--- systemd-clean/src/shared/common-signal.c
+++ systemd-workdir/src/shared/common-signal.c
@@ -66,10 +66,10 @@
                         break;
                 }
 
-                if (malloc_info(0, f) < 0) {
-                        log_error_errno(errno, "Failed to invoke malloc_info(): %m");
-                        break;
-                }
+                // if (malloc_info(0, f) < 0) {
+                //         log_error_errno(errno, "Failed to invoke malloc_info(): %m");
+                //         break;
+                // }
 
                 (void) memstream_dump(LOG_INFO, &m);
                 break;
diff -urN --no-dereference systemd-clean/src/shared/dev-setup.c systemd-workdir/src/shared/dev-setup.c
--- systemd-clean/src/shared/dev-setup.c
+++ systemd-workdir/src/shared/dev-setup.c
@@ -3,6 +3,7 @@
 #include <errno.h>
 #include <stdlib.h>
 #include <unistd.h>
+#include <sys/file.h>
 
 #include "alloc-util.h"
 #include "dev-setup.h"
