diff -urN --no-dereference xfconf-clean/configure xfconf-workdir/configure
--- xfconf-clean/configure
+++ xfconf-workdir/configure
@@ -18207,11 +18207,11 @@
     INTROSPECTION_GIRDIR=
     INTROSPECTION_TYPELIBDIR=
     if test "x$found_introspection" = "xyes"; then
-       INTROSPECTION_SCANNER=`$PKG_CONFIG --variable=g_ir_scanner gobject-introspection-1.0`
-       INTROSPECTION_COMPILER=`$PKG_CONFIG --variable=g_ir_compiler gobject-introspection-1.0`
-       INTROSPECTION_GENERATE=`$PKG_CONFIG --variable=g_ir_generate gobject-introspection-1.0`
-       INTROSPECTION_GIRDIR=`$PKG_CONFIG --variable=girdir gobject-introspection-1.0`
-       INTROSPECTION_TYPELIBDIR="$($PKG_CONFIG --variable=typelibdir gobject-introspection-1.0)"
+       INTROSPECTION_SCANNER=$INTROSPECTION_SCANNER_ENV
+       INTROSPECTION_COMPILER=$INTROSPECTION_COMPILER_ENV
+       INTROSPECTION_GENERATE=$INTROSPECTION_GENERATE_ENV
+       INTROSPECTION_GIRDIR=$INTROSPECTION_GIRDIR_ENV
+       INTROSPECTION_TYPELIBDIR=$INTROSPECTION_TYPELIBDIR_ENV
        INTROSPECTION_CFLAGS=`$PKG_CONFIG --cflags gobject-introspection-1.0`
        INTROSPECTION_LIBS=`$PKG_CONFIG --libs gobject-introspection-1.0`
        INTROSPECTION_MAKEFILE=`$PKG_CONFIG --variable=datadir gobject-introspection-1.0`/gobject-introspection-1.0/Makefile.introspection
@@ -18480,12 +18480,12 @@
   case $enable_vala in #(
   yes) :
 
-      VAPIGEN=`$PKG_CONFIG --variable=vapigen $vapigen_pkg_name`
-      VAPIGEN_MAKEFILE=`$PKG_CONFIG --variable=datadir $vapigen_pkg_name`/vala/Makefile.vapigen
+      VAPIGEN=$VAPIGEN
+      VAPIGEN_MAKEFILE=$VALA_MAKEFILE
       if test "x" = "x"
 then :
 
-          VAPIGEN_VAPIDIR=`$PKG_CONFIG --variable=vapidir $vapigen_pkg_name`
+          VAPIGEN_VAPIDIR=$VALADIR
 
 else $as_nop
 
diff -urN --no-dereference xfconf-clean/xfconf/Makefile.in xfconf-workdir/xfconf/Makefile.in
--- xfconf-clean/xfconf/Makefile.in
+++ xfconf-workdir/xfconf/Makefile.in
@@ -690,7 +690,6 @@
 @HAVE_INTROSPECTION_TRUE@typelib_DATA = $(INTROSPECTION_GIRS:.gir=.typelib)
 @HAVE_INTROSPECTION_TRUE@DISTCLEANFILES = $(gir_DATA) $(typelib_DATA)
 @ENABLE_VAPIGEN_TRUE@VAPIGEN_VAPIS = libxfconf-0.vapi
-@ENABLE_VAPIGEN_TRUE@VAPIGEN_GIRDIRS = $(prefix)/share/gir-1.0
 @ENABLE_VAPIGEN_TRUE@libxfconf_0_vapi_DEPS = gobject-2.0 glib-2.0 gio-2.0
 @ENABLE_VAPIGEN_TRUE@libxfconf_0_vapi_FILES = Xfconf-0.gir
 @ENABLE_VAPIGEN_TRUE@vapidir = $(datadir)/vala/vapi
